### DISCUSSION TITLE
Issue related to handling of NaN values in car_sales_extended_missing_data

### DISCUSSION THREAD

**Discussion by ubaidmundewadi** [mrdbourke/zero-to-mastery-ml#139] [Q&A]
Hello,

The doubt is related to video 114 in the your Data Science and Machine Learning Course

I'm using the sklearn version 1.5.1 and as mentioned in the notes the sklearn handles NaN values automatically not need to impute or delete it but after transforming the categorical data into numerical data for X when I'm trying to fit the model it's throwing an error
<img width="362" height="63" alt="image" src="https://github.com/user-attachments/assets/f9da428f-3276-4c3d-96d2-819599c3fd7b" />

Later I tried to remove NaN in "Price"column by running the below command
`car_sales_missing = car_sales_missing.dropna(subset=["Price"])`

I'm getting an error message as below
<img width="1259" height="132" alt="image" src="https://github.com/user-attachments/assets/e4e9562d-646f-4abd-ae3b-2f3900c8d799" />

Also in the video no 114 at 11:08 why we are tranforming the complete car_sales_missing data not just the X as the y (Price) is already numerical. I didn't get why we transforming the complete data

Your guidance on this topic will be greatly appreciated

Regards,
Ubaid Mundewadi

**Comment by RubensZimbres** [comment]
The core issue is order of operations: you must handle missing values in your target (y) first since no model can train on NaN labels, then handle missing values in your features (X) separately

The cleanest solution is a Scikit-Learn Pipeline that drops rows with missing targets upfront, then imputes categorical features (fill with "missing") and numeric features (fill with mean) before encoding and training. This keeps preprocessing consistent between training and prediction.

Try this code:

```
from sklearn.ensemble import RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# 1. Drop rows where target is missing, then split
data = car_sales_missing.dropna(subset=["Price"])
X = data.drop("Price", axis=1)
y = data["Price"]

# 2. Build preprocessing + model pipeline
preprocessor = ColumnTransformer([
    ("cat", Pipeline([
        ("imputer", SimpleImputer(strategy="constant", fill_value="missing")),
        ("onehot", OneHotEncoder(handle_unknown="ignore"))
    ]), ["Make", "Colour"]),
    ("num", SimpleImputer(strategy="mean"), ["Odometer (KM)", "Doors"])
])

model = Pipeline([("preprocessor", preprocessor), ("regressor", RandomForestRegressor())])
model.fit(X_train, y_train)
```
